src/hotspot/share/opto/compile.cpp

```diff
@@ -2204,6 +2209,12 @@ void Compile::Optimize() {
       // Try again if candidates exist and made progress
       // by removing some allocations and/or locks.
     } while (progress);
+
+#ifndef PRODUCT
+    Atomic::add(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr + total_scalar_replaced);
+    Atomic::add(&ConnectionGraph::_arg_escape_counter, _local_arg_escape_ctr);
+    Atomic::add(&ConnectionGraph::_global_escape_counter, _local_global_escape_ctr);
+#endif
```

Comment:
Thanks for the review @vnkozlov .

> These should be done inside ConnectionGraph - don't expose EA counters to an other class. You can use a static method in ConnectionGraph to do that.

I also didn't like this. The iterative EA and not having the last `congraph` available after the loop complicated a bit the logic here. We'll work to improve this code.

> _no_escape_counter, _local_no_escape_ctr + total_scalar_replaced is wrong. You are doubling number because total_scalar_replaced is part of _local_no_escape_ctr

I see your point here. As @navyxliu mentioned, the "local_scalar_replaced" is used as an adjustment to keep record of the objects scalar replaced in **previous iterations**. However, this approach incorrectly handles numbers in the **last iteration**. In the last iteration _local_no_escape_ctr is always correct and we don't need to  add the objects SR in that iteration.

Reply:
I see. this patch counts the scalarized objects twice in last iteration. need to subtract `mexp._local_scalar_replaced` .

Reply:
@vnkozlov Thank you for the feedback! I removed the complicated iterative feedback fix and simply left a comment in the output. My example output above is updated. I added the PrintOptoStatistics flag where it felt necessary. I also pushed my commits incrementally as opposed to squashing them and force pushing as @navyxliu suggested.